pkg/maintainer/bitcoin_difficulty.go

```diff
@@ -0,0 +1,347 @@
+// difficulty chain is up-to-date or there are not enough headers in the new
+// Bitcoin epoch), it returns false.
+func (bdm *BitcoinDifficultyMaintainer) proveNextEpoch(ctx context.Context) (
+	bool,
+	error,
+) {
+	// The height of the Bitcoin blockchain.
+	currentBlockHeight, err := bdm.btcChain.GetLatestBlockHeight()
+	if err != nil {
+		return false, fmt.Errorf(
+			"failed to get latest block height: [%w]",
+			err,
+		)
+	}
+
+	// The current epoch proven in the Bitcoin difficulty chain.
+	currentEpoch, err := bdm.chain.CurrentEpoch()
+	if err != nil {
+		return false, fmt.Errorf(
+			"failed to get current epoch: [%w]",
+			err,
+		)
+	}
+
+	// The number of blocks required for each side of a retarget proof.
+	proofLength, err := bdm.chain.ProofLength()
+	if err != nil {
+		return false, fmt.Errorf(
+			"failed to get proof length: [%w]",
+			err,
+		)
+	}
+
+	// The new epoch to be proven in the Bitcoin difficulty chain.
+	newEpoch := uint(currentEpoch) + 1
+
+	// Height of the first block of the new epoch.
+	newEpochHeight := newEpoch * bitcoinDifficultyEpochLength
+
+	// The range of block headers to be pulled from the Bitcoin chain should
+	// start `proofLength` blocks before the first block of the new difficulty
+	// epoch and end `proofLength`-1 after it.
+	// For example, if the new epoch begins at block 522144 and `proofLength`
+	// is 3, then the range should be [522141, 522146]:
+	// 522141 <- old difficulty target
+	// 522142 <- old difficulty target
+	// 522143 <- old difficulty target
+	// << difficulty retarget >>
+	// 522144 <- new difficulty target (first block of the new epoch)
+	// 522145 <- new difficulty target
+	// 522146 <- new difficulty target
+	firstBlockHeaderHeight := newEpochHeight - uint(proofLength)
+	lastBlockHeaderHeight := newEpochHeight + uint(proofLength) - 1
+
+	// The required range of block headers can be pulled from the Bitcoin
+	// blockchain only if the blockchain height is equal to or greater than
+	// the end of the range.
+	if currentBlockHeight >= lastBlockHeaderHeight {
+		headers, err := bdm.getBlockHeaders(
+			firstBlockHeaderHeight,
+			lastBlockHeaderHeight,
+		)
+		if err != nil {
+			return false, fmt.Errorf(
+				"failed to get block headers from Bitcoin chain: [%w]",
+				err,
+			)
+		}
+
+		if err := bdm.chain.Retarget(headers); err != nil {
+			return false, fmt.Errorf(
+				"failed to submit block headers from range [%d:%d] to "+
+					"the Bitcoin difficulty chain: [%w]",
+				firstBlockHeaderHeight,
+				lastBlockHeaderHeight,
+				err,
+			)
+		}
+
+		if err := bdm.waitForCurrentEpochUpdate(ctx, uint64(newEpoch)); err != nil {
+			return false, fmt.Errorf(
+				"error while waiting for current Bitcoin difficulty epoch "+
+					"update: [%w]",
+				err,
+			)
+		}
+
+		logger.Infof(
+			"successfully submitted block headers [%d:%d] to the Bitcoin "+
+				"difficulty chain. The current proven epoch is %d.",
```

Comment:
```suggestion
				"difficulty chain; the current proven epoch is [%d]",
```

Reply:
Done in 5fa25eb824348e461e52ca1279373fe17a376f12.